godot-core/src/builtin/signal/futures.rs

```diff
@@ -0,0 +1,313 @@
+#[cfg_attr(feature = "trace", derive(Default))]
+pub struct SignalFutureResolver<R> {
+    data: Arc<Mutex<SignalFutureData<R>>>,
+}
+
+impl<R> Clone for SignalFutureResolver<R> {
+    fn clone(&self) -> Self {
+        Self {
+            data: self.data.clone(),
+        }
+    }
+}
+
+impl<R> SignalFutureResolver<R> {
+    fn new(data: Arc<Mutex<SignalFutureData<R>>>) -> Self {
+        Self { data }
+    }
+}
+
+impl<R> std::hash::Hash for SignalFutureResolver<R> {
+    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
+        state.write_usize(Arc::as_ptr(&self.data) as usize);
+    }
+}
+
+impl<R> PartialEq for SignalFutureResolver<R> {
+    fn eq(&self, other: &Self) -> bool {
+        Arc::ptr_eq(&self.data, &other.data)
+    }
+}
+
+impl<R: ParamTuple + Sync + Send> RustCallable for SignalFutureResolver<R> {
+    fn invoke(&mut self, args: &[&Variant]) -> Result<Variant, ()> {
+        let mut data = self.data.lock().unwrap();
+        // We no loger need the waker after we resolved. If the future get's polled again we also get a new waker.
+        let waker = data.waker.take();
+
+        data.state = TrySignalFutureState::Ready(R::from_variant_array(args));
+        drop(data);
+
+        if let Some(waker) = waker {
+            waker.wake();
+        }
+
+        Ok(Variant::nil())
+    }
+}
+
+impl<R> Display for SignalFutureResolver<R> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "GuaranteedSignalFutureResolver::<{}>",
+            std::any::type_name::<R>()
+        )
+    }
+}
+
+// this resolver will resolve the future when it's being dropped (i.e. the engine removes all connected signal callables). This is very unusual.
+impl<R> Drop for SignalFutureResolver<R> {
+    fn drop(&mut self) {
+        let mut data = self.data.lock().unwrap();
+
+        if !matches!(data.state, TrySignalFutureState::Pending) {
+            return;
+        }
+
+        data.state = TrySignalFutureState::Dead;
+
+        if let Some(ref waker) = data.waker {
+            waker.wake_by_ref();
+        }
+    }
+}
+
+#[derive(Default)]
+enum TrySignalFutureState<T> {
+    #[default]
+    Pending,
+    Ready(T),
+    Dead,
+    Dropped,
+}
+
+impl<T> TrySignalFutureState<T> {
+    fn take(&mut self) -> Self {
+        let new_value = match self {
+            Self::Pending => Self::Pending,
+            Self::Ready(_) | Self::Dead => Self::Dead,
+            Self::Dropped => Self::Dropped,
+        };
+
+        std::mem::replace(self, new_value)
+    }
+}
+
+/// A future that tries to resolve as soon as the provided Godot signal was emitted.
+///
+/// The future might resolve to an error if the signal object is freed before the signal is emitted.
+pub struct TrySignalFuture<R: ParamTuple + Sync + Send> {
+    data: Arc<Mutex<SignalFutureData<R>>>,
+    callable: SignalFutureResolver<R>,
+    signal: Signal,
+}
+
+impl<R: ParamTuple + Sync + Send> TrySignalFuture<R> {
```

Comment:
You're not always consistent with the generic argument, sometimes it's called `R` and sometimes `Ps`.

What does it represent exactly, which identifier is better fitted? It can also help to mention it in the RustDoc of the struct.

Reply:
Right, I think it should always be `R` that is what I initially used and stands for `Return`. `Ps` accidentally crept in when I adopted the `ParamTuple` from the `TypedSignal`.